pkg/services/ngalert/notifier/alertmanager.go

```diff
@@ -70,72 +87,110 @@ func (am *Alertmanager) Init() (err error) {
 		return errors.Wrap(err, "unable to initialize the silencing component of alerting")
 	}
 
+	am.alerts, err = NewAlertProvider(nil, am.marker)
+	if err != nil {
+		return errors.Wrap(err, "unable to initialize the alert provider component of alerting")
+	}
+
 	return nil
 }
 
 func (am *Alertmanager) Run(ctx context.Context) error {
 	// Make sure dispatcher starts. We can tolerate future reload failures.
-	if err := am.ReloadConfigFromDatabase(); err != nil {
+	if err := am.SyncAndApplyConfigFromDatabase(); err != nil && errors.Is(err, store.ErrNoAlertmanagerConfiguration) {
 		return err
 	}
+
 	for {
 		select {
 		case <-ctx.Done():
 			am.StopAndWait()
 			return nil
 		case <-time.After(1 * time.Minute):
 			// TODO: once we have a check to skip reload on same config, uncomment this.
-			//if err := am.ReloadConfigFromDatabase(); err != nil {
-			//	am.logger.Error("failed to sync config from database", "error", err)
+			//if err := am.SyncAndApplyConfigFromDatabase(); err != nil {
+			//	if err == store.ErrNoAlertmanagerConfiguration {
+			//		am.logger.Warn(errors.Wrap(err, "unable to sync configuration").Error())
+			//	}
+			//	am.logger.Error(errors.Wrap(err, "unable to sync configuration").Error())
 			//}
 		}
 	}
 }
 
+// AddMigration runs the database migrations as the service starts.
+func (am *Alertmanager) AddMigration(mg *migrator.Migrator) {
+	alertmanagerConfigurationMigration(mg)
+}
+
```

Comment:
Am I missing something or is it not being not called anywhere now after you moved `alertmanagerConfigurationMigration`? Also we could just expose `alertmanagerConfigurationMigration` and not needed to go through Alertmanager.

Reply:
Not sure I get this statement. `AddMigration` is part of Grafana's service interface, it is run by the DI framework to call this method as we initilize the service:

```
// DatabaseMigrator allows the caller to add migrations to
// the migrator passed as argument
type DatabaseMigrator interface {
	// AddMigrations allows the service to add migrations to
	// the database migrator.
	AddMigration(mg *migrator.Migrator)
}
```

Reply:
> Not sure I get this statement. AddMigration is part of Grafana's service interface, it is run by the DI framework to call this method as we initilize the service:

exactly! check also [this](https://github.com/grafana/grafana/blob/master/contribute/architecture/backend/database.md#implement-databasemigrator)